File: app/src/main/java/fr/smarquis/applinks/Referrer.java
package fr.smarquis.applinks;


import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

final class Referrer {

    private static final String PREFS_NAME = "referrer";
    private static final String FIRST_LAUNCH = "FIRST_LAUNCH";
    private static final String REFERRER_DATE = "REFERRER_DATE";
    private static final String REFERRER_DATA = "REFERRER_DATA";
    private static final String DISPLAYED = "DISPLAYED";

    private Referrer() {
    }

    private static SharedPreferences prefs(Context context) {
        return context.getApplicationContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    static void setFirstLaunch(Context context) {
        SharedPreferences sp = prefs(context);
        if (!sp.contains(FIRST_LAUNCH)) {
            sp.edit().putLong(FIRST_LAUNCH, System.currentTimeMillis()).apply();
        }
    }

    static String getFirstLaunch(Context context) {
        long time = prefs(context).getLong(FIRST_LAUNCH, 0);
        if (time > 0) {
            return formatDateTime(time);
        }
        return null;
    }

    static boolean isAvailable(Context context) {
        return prefs(context).contains(REFERRER_DATE);
    }

    static void setDisplayed(Context context) {
        prefs(context).edit().putBoolean(DISPLAYED, true).apply();
    }

    static boolean hasBeenDisplayed(Context context) {
        return prefs(context).getBoolean(DISPLAYED, false);
    }

    static void setReferrer(Context context, String data) {
        SharedPreferences sp = prefs(context);
        sp.edit().putLong(REFERRER_DATE, System.currentTimeMillis())
                .putString(REFERRER_DATA, data)
                .apply();
    }

    @Nullable
    static String getDate(Context context) {
        long time = prefs(context).getLong(REFERRER_DATE, 0);
        if (time > 0) {
            return formatDateTime(time);
        }
        return null;
    }

    @NonNull
    private static String formatDateTime(long time) {
        Date date = new Date(time);
        return DateFormat.getDateInstance().format(date) + " - " + new SimpleDateFormat("HH:mm:ss.SSS", Locale.getDefault()).format(date);
    }

    @Nullable
    static String getRawData(Context context) {
        return prefs(context).getString(REFERRER_DATA, null);
    }

    @Nullable
    static String getDecodedData(Context context) {
        String raw = getRawData(context);
        if (raw == null) {
            return null;
        }

        String first = urlDecode(raw);
        if (first == null || first.equals(raw)) {
            return null;
        }
        String second = urlDecode(first);
        if (second == null || second.equals(raw)) {
            return first;
        }
        return second;
    }

    @Nullable
    private static String urlDecode(@NonNull String input) {
        try {
            return URLDecoder.decode(input, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            return null;
        }
    }
}

File: app/src/main/java/fr/smarquis/applinks/Referrer.kt
package fr.smarquis.applinks

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.android.installreferrer.api.ReferrerDetails

internal object Referrer {

    private const val PREFS_NAME = "referrer"
    private const val KEY_FIRST_LAUNCH = "FIRST_LAUNCH"
    private const val KEY_INSTALL_REFERRER = "install_referrer"
    private const val KEY_REFERRER_CLICK_TIMESTAMP = "referrer_click_timestamp_seconds"
    private const val KEY_INSTALL_BEGIN_TIMESTAMP = "install_begin_timestamp_seconds"
    private const val KEY_GOOGLE_PLAY_INSTANT = "google_play_instant"
    private const val KEY_REFERRER_CLICK_TIMESTAMP_SERVER = "referrer_click_timestamp_server_seconds"
    private const val KEY_INSTALL_BEGIN_TIMESTAMP_SERVER = "install_begin_timestamp_server_seconds"
    private const val KEY_INSTALL_VERSION = "install_version"

    data class LocalReferrerDetails(
        val firstLaunch: Long,
        val installReferrer: String?,
        val referrerClickTimestampSeconds: Long,
        val installBeginTimestampSeconds: Long,
        val googlePlayInstantParam: Boolean,
        val referrerClickTimestampServerSeconds: Long,
        val installBeginTimestampServerSeconds: Long,
        val installVersion: String?,
    )


    private val INTENT = Intent("custom-action-local-broadcast")
    val INTENT_FILTER = IntentFilter(INTENT.action)

    private fun prefs(context: Context): SharedPreferences = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun init(context: Context, details: ReferrerDetails) {
        val prefs = prefs(context)
        if (prefs.all.isNotEmpty()) return
        prefs.edit().apply {
            putLong(KEY_FIRST_LAUNCH, System.currentTimeMillis())
            putString(KEY_INSTALL_REFERRER, details.installReferrer)
            putLong(KEY_REFERRER_CLICK_TIMESTAMP, details.referrerClickTimestampSeconds)
            putLong(KEY_INSTALL_BEGIN_TIMESTAMP, details.installBeginTimestampSeconds)
            putBoolean(KEY_GOOGLE_PLAY_INSTANT, details.googlePlayInstantParam)
            putLong(KEY_REFERRER_CLICK_TIMESTAMP_SERVER, details.referrerClickTimestampServerSeconds)
            putLong(KEY_INSTALL_BEGIN_TIMESTAMP_SERVER, details.installBeginTimestampServerSeconds)
            putString(KEY_INSTALL_VERSION, details.installVersion)
        }.apply()
        LocalBroadcastManager.getInstance(context).sendBroadcast(INTENT)
    }

    fun isAvailable(context: Context): Boolean = prefs(context).all.isNotEmpty()

    fun get(context: Context) = prefs(context).run {
        LocalReferrerDetails(
            firstLaunch = getLong(KEY_FIRST_LAUNCH, 0),
            installReferrer = getString(KEY_INSTALL_REFERRER, null),
            referrerClickTimestampSeconds = getLong(KEY_REFERRER_CLICK_TIMESTAMP, 0),
            installBeginTimestampSeconds = getLong(KEY_INSTALL_BEGIN_TIMESTAMP, 0),
            googlePlayInstantParam = getBoolean(KEY_GOOGLE_PLAY_INSTANT, false),
            referrerClickTimestampServerSeconds = getLong(KEY_REFERRER_CLICK_TIMESTAMP_SERVER, 0),
            installBeginTimestampServerSeconds = getLong(KEY_INSTALL_BEGIN_TIMESTAMP_SERVER, 0),
            installVersion = getString(KEY_INSTALL_VERSION, null),
        )
    }

}